extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -0,0 +1,243 @@
+          }
+        },
+      };
+      this.vsync_.run(task, {});
+    }
+  }
+
+  /**
+   * Applies `newValue` to the element bound in `binding`.
+   * @param {!BindingDef} binding
+   * @param {(Object|Array|string|number|boolean|null)} newValue
+   * @private
+   */
+  applyBinding_(binding, newValue) {
+    const property = binding.property;
+    const element = binding.element;
+
+    // TODO: Support arrays for classes and objects for attributes.
+
+    if (property === 'text') {
+      element.textContent = newValue;
+    } else if (property === 'class') {
+      element.classList = newValue;
+    } else {
+      if (newValue === true) {
+        element.setAttribute(property, '');
+      } else if (newValue === false) {
+        element.removeAttribute(property);
+      } else {
+        const oldValue = element.getAttribute(property);
+        const sanitizedValue = this.sanitizeAttribute_(newValue);
+
+        if (oldValue === sanitizedValue) {
+          return;
+        }
+
+        element.setAttribute(property, sanitizedValue);
+
+        // Update internal state for AMP elements.
+        if (element.classList.contains('-amp-element')) {
+          const resources = element.getResources();
+          if (property === 'width') {
+            resources.changeSize(element, undefined, sanitizedValue);
+          } else if (property === 'height') {
+            resources.changeSize(element, sanitizedValue, undefined);
+          }
+          element.attributeChangedCallback(property, oldValue, sanitizedValue);
+        }
+      }
+    }
+  }
+
+  /**
+   * If the current value of `binding` equals `expectedValue`, returns true.
+   * Otherwise, returns false.
+   * @param {!BindingDef} binding
+   * @param {?(Object|Array|string|number|boolean)} expectedValue
+   * @private
+   */
+  verifyBinding_(binding, expectedValue) {
+    const property = binding.property;
+    const element = binding.element;
+
+    // TODO: Support arrays for classes and objects for attributes.
+
+    let initialValue;
+    if (property === 'text') {
+      initialValue = element.textContent;
+    } else if (property === 'class') {
+      initialValue = element.classList;
+    } else {
+      const attribute = element.getAttribute(property);
+      if (typeof expectedValue === 'boolean') {
+        initialValue = !!attribute;
+      } else {
+        initialValue = attribute;
+      }
+    }
+
+    if (initialValue != expectedValue) {
+      user().error(TAG_,
+        `<${element.tagName}> element [${property}] binding ` +
+        `default value (${initialValue}) does not match first expression ` +
+        `result (${expectedValue}).`);
+    }
+  }
+
+  /**
+   * Sanitizes unsafe protocols in attributes, e.g. "javascript:".
+   * @param {(Object|Array|string|number|boolean|null)} value
+   * @return {string}
+   * @private
+   */
+  sanitizeAttribute_(value) {
+    if (typeof value === 'string') {
+      const split = value.split(':');
```

Comment:
This'll have to pass security review, but might be able specialize our protocol search a bit more:

```js
const protocol = `/^([A-Za-z]+([A-Za-z+.-])*):/`.match(value);
if (protocol && !this.protocolWhitelist_[protocol[0]]) {
  // unsafe
}
// safe
```